File: test/io_test.cc
/** Copyright 2021 Alibaba Group Holding Limited.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include <bitset>
#include <iostream>

#include "common/util/logging.h"
#include "common/util/uuid.h"
#include "io/io/io_factory.h"

void ReadLines(std::string const& path_to_read) {
  auto io = vineyard::IOFactory::CreateIOAdaptor(path_to_read, nullptr);
  VINEYARD_CHECK_OK(io->Open());
  size_t lineno = 1;
  std::string line;
  while (true) {
    auto s = io->ReadLine(line);
    if (s.ok()) {
      LOG(INFO) << "line " << (lineno++) << ": '" << line << "' of length "
                << line.length();
    } else {
      LOG(ERROR) << "failed (or finished): " << s.ToString();
      break;
    }
  }
}

void ReadTable(std::string const& path_to_read) {
  constexpr int total_parts = 4;
  for (int index = 0; index < total_parts; ++index) {
    auto io = vineyard::IOFactory::CreateIOAdaptor(path_to_read, nullptr);
    VINEYARD_CHECK_OK(io->SetPartialRead(index, total_parts));
    VINEYARD_CHECK_OK(io->Open());

    for (auto const& item : io->GetMeta()) {
      LOG(INFO) << "item: " << item.first << " -> " << item.second;
    }

    // read as lines first
    {
      size_t lineno = 1;
      std::string line;
      while (true) {
        auto s = io->ReadLine(line);
        if (s.ok()) {
          LOG(INFO) << "line " << (lineno++) << ": '" << line << "' of length "
                    << line.length();
        } else {
          LOG(ERROR) << "failed (or finished): " << s;
          break;
        }
      }
    }

    // read as tables
    {
      std::shared_ptr<arrow::Table> table;
      LOG(INFO) << "read table: " << index << " of part " << total_parts << ": "
                << io->ReadTable(&table);
      if (table) {
        LOG(INFO) << "table: " << table->ToString();
      }
    }
  }
}

int main(int argc, char** argv) {
  if (argc < 3) {
    printf("usage ./io_test <lines or table> <path to read>");
    return 1;
  }

  std::string mode = std::string(argv[1]);
  std::string path_to_read = std::string(argv[2]);

  if (mode == "lines") {
    ReadLines(path_to_read);
  }
  if (mode == "table") {
    ReadTable(path_to_read);
  }

  LOG(INFO) << "Passed double array tests...";

  return 0;
}

File: k8s/cmd/commands/util/io_test.go
/*
* Copyright 2020-2023 Alibaba Group Holding Limited.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

	http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
package util

import (
	"os"
	"testing"
)

func Test_ReadFromFile(t *testing.T) {
	tempFileContent := ""
	tempFile, err := os.CreateTemp("", "testfile")
	if err != nil {
		t.Fatalf("Failed to create temporary file: %v", err)
	}
	defer tempFile.Close()
	if err = os.WriteFile(tempFile.Name(), []byte(tempFileContent), 0644); err != nil {
		t.Errorf("Failed to write file: %v", err)
	}

	type args struct {
		path string
	}
	tests := []struct {
		name    string
		args    args
		want    string
		wantErr bool
	}{
		{
			name: "Test case 1",
			args: args{
				path: tempFile.Name(),
			},
			want:    tempFileContent,
			wantErr: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ReadFromFile(tt.args.path)
			if (err != nil) != tt.wantErr {
				t.Errorf("ReadFromFile() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if got != tt.want {
				t.Errorf("ReadFromFile() = %v, want %v", got, tt.want)
			}
		})
	}
}

func Test_ReadFromStdin(t *testing.T) {
	type args struct {
		args []string
	}
	tests := []struct {
		name    string
		args    args
		want    string
		wantErr bool
	}{
		{
			name: "Test case 1",
			args: args{
				args: []string{"-"},
			},
			want:    "Test input",
			wantErr: false,
		},
	}

	expected := "Test input"
	r, w, _ := os.Pipe()
	os.Stdin = r
	go func() {
		defer w.Close()
		if _, err := w.Write([]byte(expected)); err != nil {
			t.Errorf("Failed to write file: %v", err)
		}
	}()

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ReadFromStdin(tt.args.args)
			if (err != nil) != tt.wantErr {
				t.Errorf("ReadFromStdin() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if got != tt.want {
				t.Errorf("ReadFromStdin() = %v, want %v", got, tt.want)
			}
		})
	}
}

func Test_ReadJsonFromStdin(t *testing.T) {
	type args struct {
		args []string
	}
	tests := []struct {
		name    string
		args    args
		want    string
		wantErr bool
	}{
		{
			name: "Test case 1",
			args: args{
				args: []string{"-"},
			},
			want:    `{"test": "input"}`,
			wantErr: false,
		},
	}

	expected := `{"test": "input"}`
	r, w, _ := os.Pipe()
	os.Stdin = r
	go func() {
		defer w.Close()
		if _, err := w.Write([]byte(expected)); err != nil {
			t.Errorf("Failed to write file: %v", err)
		}
	}()

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ReadJsonFromStdin(tt.args.args)
			if (err != nil) != tt.wantErr {
				t.Errorf("ReadJsonFromStdin() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if got != tt.want {
				t.Errorf("ReadJsonFromStdin() = %v, want %v", got, tt.want)
			}
		})
	}
}
